File: JavaScript/7_DynamicProgramming/1_FibanocciSeries.js
// Basic Recursion for Fib Series
//[1, 1, 2, 3 ...]
//Time Complexity => O(2^n)   =>  very bad
const Fib_Basic = (n) => {
  if (n <= 2) return 1;
  return Fib_Basic(n - 1) + Fib_Basic(n - 2);
};

// Fib Function with Memorisation
//Time Complexity => O(n)   =>  Good

const Fib_Memorzation = (n, memo = [0, 1, 1]) => {
  if (memo[n] !== undefined) return memo[n];
  memo[n] = Fib_Memorzation(n - 1, memo) + Fib_Memorzation(n - 2, memo);
  return memo[n];
};

// Fib Function with Tabularization
//Time Complexity => O(n)   =>  Good
//Better than Memorization when comparing Space Complexity

const Fib_Tabularization = (n) => {
  let fibList = [0, 1, 1];
  for (let i = 3; i <= n; i++) {
    fibList[i] = fibList[i - 1] + fibList[i - 2];
  }
  return fibList[n];
};

let val = 40;
console.log(Fib_Memorzation(val));
console.log(Fib_Tabularization(val));
console.log(Fib_Basic(val));

File: Python/7_DynamicProgramming/1_FibanocciSeries.py


def Fib_Basic(num):
	if num <= 2: return 1
	return Fib_Basic(num-1) + Fib_Basic(num-2)


# this takes more time than FIB_memo
def Fib_Memorisation(num):
	fib_dict = {
	1:1,
	2:1
	}

	def FIB(num):
		if num in fib_dict:
			return fib_dict[num]

		val = FIB(num - 1) + FIB(num - 2)
		fib_dict[num] = val
		return val
	return FIB(num)


def Fib_memo(num, memo = {1:1, 2:1}):
	if num in memo: return memo[num]

	memo[num] = Fib_memo(num-1, memo) + Fib_memo(num-2, memo)
	return memo[num]

def Fib_Tabular(num):
	fib_lst = [0, 1, 1]

	for i in range(3,num + 1):
		fib_lst.append(fib_lst[i-1] + fib_lst[i-2])
	return fib_lst[num]